src/Bicep.Core/Emit/TemplateWriter.cs

```diff
@@ -197,51 +177,108 @@ private void EmitParametersIfPresent(ExpressionEmitter emitter, ImmutableArray<D
             });
         }
 
-        private ObjectExpression AddDecoratorsToBody(DecorableSyntax decorated, ObjectExpression input, TypeSymbol targetType)
+        private ObjectExpression ApplyTypeModifiers(TypeDeclaringExpression expression, ObjectExpression input)
         {
-            var result = input;
-            foreach (var decoratorSyntax in decorated.Decorators.Reverse())
-            {
-                var symbol = this.Context.SemanticModel.GetSymbolInfo(decoratorSyntax.Expression);
+            List<(Expression source, Func<ObjectExpression, ObjectExpression> modifyFunc)> modifications = new();
 
-                if (symbol is FunctionSymbol decoratorSymbol &&
-                    decoratorSymbol.DeclaringObject is NamespaceType namespaceType &&
-                    DecoratorsToEmitAsResourceProperties.Contains(decoratorSymbol.Name))
+            if (expression.Secure is {} secure)
+            {
+                modifications.Add((secure, result =>
```

Comment:
I think we can extract a helper (something like `AddIfExpressionNotNull`) to remove the need for expression null checks in this method.